### PR TITLE
Faster gradient rules for {numpy,scipy}.linalg.solve

### LINKED ISSUE #1747: Direct gradient rules to speed-up numpy.linalg.solve
It can be [significantly more efficient](https://www.semanticscholar.org/paper/Structured-higher-order-algorithmic-differentiation-Walter/e52d834dac4fdfdc4328105e29ae657c3fa51467), both in runtime and memory, to directly define derivative rules for higher level linear algebra operations like `solve` rather than the constituent operations (factorization and triangular solve).

For large matrices (e.g., 500x500 on CPUs), my microbenchmark shows that we can get a 3-4x speed-up for general purpose and symmetric solves:
```python
from functools import partial
import jax.scipy as jsp
from jax import lax
import jax.numpy as np
import numpy as onp
import jax

def positive_definite_solve(a, b):
  factors = jsp.linalg.cho_factor(a)
  def solve(matvec, x):
    return jsp.linalg.cho_solve(factors, x)
  matvec = partial(np.dot, a)
  return lax.custom_linear_solve(matvec, b, solve, symmetric=True)

def linear_solve(a, b):
  a_factors = jsp.linalg.lu_factor(a)
  def solve(matvec, x):
    return jsp.linalg.lu_solve(a_factors, x)
  def transpose_solve(vecmat, x):
    return jsp.linalg.lu_solve(a_factors, x, trans=1)
  matvec = partial(np.dot, a)
  return lax.custom_linear_solve(matvec, b, solve, transpose_solve)

def loss(solve):
  def f(a, b):
    return solve(a, b).sum()
  return f

rs = onp.random.RandomState(0)
a = rs.randn(500, 500)
a = jax.device_put(a.T @ a + 0.1 * np.eye(500))
b = jax.device_put(rs.randn(500))

# general purpose solve
# current
grad = jax.jit(jax.grad(loss(np.linalg.solve)))
%timeit jax.device_get(grad(a, b))
# 33.8 ms per loop
# new
grad = jax.jit(jax.grad(loss(linear_solve)))
%timeit jax.device_get(grad(a, b))
# 10.1 ms per loop

# positive definite solve
# current
grad = jax.jit(jax.grad(loss(partial(jsp.linalg.solve, sym_pos=True))))
%timeit jax.device_get(grad(a, b))
# 23.7 ms per loop
# new
grad = jax.jit(jax.grad(loss(positive_definite_solve)))
%timeit jax.device_get(grad(a, b))
# 4.8 ms per loop
```

Unfortunately, we can't just use these prototype implementations internally in JAX, for two reasons:
1. ~~`custom_linear_solve` (like custom transforms in general) doesn't work with batching yet (https://github.com/google/jax/issues/1249).~~ This was solved by #2099.
2. ~~We do an extra optimization in `triangular_solve_jvp_rule_a` for the case of solving many right-hand-sides at the same time with the same left-hand side (https://github.com/google/jax/pull/1466). This new gradient rule here doesn't handle this yet.~~ Update: in practice, I don't think this optimization actually matters -- it's the difference between `n*m*m+m*m*m` time vs `2*m*m*m` time.
3. We need to support multiple right-hand-side arguments. After https://github.com/google/jax/pull/2138, we'll be able to do this by batching `custom_linear_solve`.

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #1747

The implicit function theorem (via `lax.custom_linear_solve`) lets us _directly_ define gradients for linear solves, in contrast to the current implementations of gradient for `solve` which rely upon differentiating matrix factorization.

In **theory**, JVPs of `cholesky` and `lu` involve the equivalent of ~3 dense matrix-matrix multiplications, which makes them rather expensive: time `O(n**3)`. In contrast, with `custom_linear_solve` we don't need to differentiate the factorization. The JVP and VJP rules for linear solve (for a single right-hand-side vector) now only use matrix-vector products and triangular solves, which is time `O(n**2)`. We should also have reduced memory usage, because we don't need to save any intermediate outputs.

In **practice**, these new gradient rules seem to make linear solves with large arrays ~3x faster, e.g.,

    from functools import partial
    import jax.scipy as jsp
    from jax import lax
    import jax.numpy as np
    import numpy as onp
    import jax

    def loss(solve):
      def f(a, b):
        return solve(a, b).sum()
      return f

    rs = onp.random.RandomState(0)
    N = 500
    K = 1
    a = rs.randn(N, N)
    a = jax.device_put(a.T @ a + 0.1 * np.eye(N))
    b = jax.device_put(rs.randn(N, K))

    # general matrix solve
    grad = jax.jit(jax.grad(loss(np.linalg.solve)))
    grad(a, b).block_until_ready()
    %timeit grad(a, b).block_until_ready()
    # N=500, K=1: 11.4 ms -> 3.63 ms

    # positive definite solve
    grad = jax.jit(jax.grad(loss(partial(jsp.linalg.solve, sym_pos=True))))
    grad(a, b).block_until_ready()
    %timeit grad(a, b).block_until_ready()
    # N=500, K=1: 9.22 ms -> 2.83 ms

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
Seems like a great improvement! Thanks, @shoyer !

**Comment by shoyer** [comment]
thanks for the review @mattjj !
